File: algorithmic-toolbox/W2/08 Last Digit of the Sum of Squares of Fibonacci Numbers/fibonacci_sum_squares.py
# python3
from sys import stdin


def get_fibonacci_last_digit(n):
    if (n < 1):
        return n

    prev = 0
    curr = 1

    for _ in range(n - 1):
        prev, curr = curr % 10, (prev + curr) % 10

    return curr % 10

def get_fibonacci_sum_squares(n):
    # Little trick:
    # Pisano period modulo 10 is 60
    vertical_side = get_fibonacci_last_digit(n % 60)
    horizontal_side = get_fibonacci_last_digit((n + 1) % 60)
    return (vertical_side * horizontal_side) % 10

if __name__ == '__main__':
    n = int(stdin.read())
    print(get_fibonacci_sum_squares(n))

File: algorithmic-toolbox/W2/08 Last Digit of the Sum of Squares of Fibonacci Numbers/FibonacciSumSquares.java
import java.util.*;

public class FibonacciSumSquares {
    private static final int PISANO_PERIOD_MOD10 = 60;

    private static int getFibonacciLastDigit(long n) {
        if (n <= 1)
            return (int) n;

        int prev = 0;
        int curr = 1;
        for (long i = 2; i <= n; i++) {
            int next = (prev + curr) % 10;
            prev = curr;
            curr = next;
        }
        return curr % 10;
    }

    private static int getFibonacciSumSquares(long n) {
        int verticalSide = getFibonacciLastDigit(n % PISANO_PERIOD_MOD10);
        int horizontalSide = getFibonacciLastDigit((n + 1) % PISANO_PERIOD_MOD10);
        int sumSquares = verticalSide * horizontalSide;
        return sumSquares % 10;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        long s = getFibonacciSumSquares(n);
        System.out.println(s);
    }
}
